From e7b7c7257069e9d641eca0b2f0ca26069cfa1bbe Mon Sep 17 00:00:00 2001
From: Martin Magr <mmagr@redhat.com>
Date: Fri, 25 Apr 2014 15:48:27 +0200
Subject: [PATCH] Use lioadm as iscsi helper on RHEL 7

On RHEL 7 it is necessary to use lioadm as iscsi_helper. This patch handles
proper configuration and appropriate package installation.

Change-Id: Ic3a41015c95f0b5f394d0c4b58dfc0eb10d5037d

Conflicts:
	manifests/backend/iscsi.pp
	spec/defines/cinder_backend_iscsi_spec.rb
---
 manifests/params.pp                      |  9 +++++++++
 manifests/volume/iscsi.pp                |  9 ++++++++-
 spec/classes/cinder_volume_iscsi_spec.rb | 22 ++++++++++++++++++++--
 3 files changed, 37 insertions(+), 3 deletions(-)

diff --git a/manifests/params.pp b/manifests/params.pp
index 56b51ed..9d5da31 100644
--- a/manifests/params.pp
+++ b/manifests/params.pp
@@ -19,6 +19,8 @@ class cinder::params {
     $tgt_package_name   = 'tgt'
     $tgt_service_name   = 'tgt'
     $ceph_init_override = '/etc/init/cinder-volume.override'
+    $iscsi_helper       = 'tgtadm'
+    $lio_package_name   = 'targetcli'
 
   } elsif($::osfamily == 'RedHat') {
 
@@ -36,6 +38,13 @@ class cinder::params {
     $tgt_package_name   = 'scsi-target-utils'
     $tgt_service_name   = 'tgtd'
     $ceph_init_override = '/etc/sysconfig/openstack-cinder-volume'
+    $lio_package_name   = 'targetcli'
+
+    if $::operatingsystem == 'RedHat' and $::operatingsystemrelease >= 7 {
+      $iscsi_helper = 'lioadm'
+    } else {
+      $iscsi_helper = 'tgtadm'
+    }
 
   } else {
     fail("unsuported osfamily ${::osfamily}, currently Debian and Redhat are the only supported platforms")
diff --git a/manifests/volume/iscsi.pp b/manifests/volume/iscsi.pp
index 4c6e05d..0322187 100644
--- a/manifests/volume/iscsi.pp
+++ b/manifests/volume/iscsi.pp
@@ -2,7 +2,7 @@
 class cinder::volume::iscsi (
   $iscsi_ip_address,
   $volume_group      = 'cinder-volumes',
-  $iscsi_helper      = 'tgtadm'
+  $iscsi_helper      = $cinder::params::iscsi_helper,
 ) {
 
   include cinder::params
@@ -38,6 +38,13 @@ class cinder::volume::iscsi (
       }
     }
 
+    'lioadm': {
+      package { 'targetcli':
+        ensure => present,
+        name   => $::cinder::params::lio_package_name,
+      }
+    }
+
     default: {
       fail("Unsupported iscsi helper: ${iscsi_helper}.")
     }
diff --git a/spec/classes/cinder_volume_iscsi_spec.rb b/spec/classes/cinder_volume_iscsi_spec.rb
index fe0a7f7..0e7a681 100644
--- a/spec/classes/cinder_volume_iscsi_spec.rb
+++ b/spec/classes/cinder_volume_iscsi_spec.rb
@@ -2,8 +2,10 @@ require 'spec_helper'
 
 describe 'cinder::volume::iscsi' do
 
-  let :req_params do
-    {:iscsi_ip_address => '127.0.0.2'}
+  let :req_params do {
+    :iscsi_ip_address => '127.0.0.2',
+    :iscsi_helper => 'tgtadm'
+  }
   end
 
   let :facts do
@@ -45,4 +47,20 @@ describe 'cinder::volume::iscsi' do
 
   end
 
+  describe 'with lioadm' do
+
+    let :params do {
+      :iscsi_ip_address => '127.0.0.2',
+      :iscsi_helper => 'lioadm'
+    }
+    end
+
+    let :facts do
+      {:osfamily => 'RedHat'}
+    end
+
+    it { should contain_package('targetcli').with_ensure('present')}
+
+  end
+
 end
-- 
1.8.3.1

